---
title: "Demultipexing and adapter trimming of 16S rRNA gene sequences of DNA from Oman samples collected in 2014"
subtitle: "Source file: OM14_seq_prep.Rmd"
author: "Daniel Nothaft"
date: "`r format(Sys.Date(), '%d %b %Y')`"
output:
  html_document: 
    df_print: paged # omit to disable paged data table output
    css: stylesheet.css # omit if no need for custom stylesheet
    number_sections: yes # change to no for unnumbered sections
    toc: yes # change to no to disable table of contents
    toc_float: true # change to false to keep toc at the top
    toc_depth: 3 # change to specify which headings to include in toc
    code_folding: show # change to hide to hide code by default
editor_options:
  chunk_output_type: inline
---

# Setup

Set knitting options
```{r knitting-options}
# global knitting options for automatic saving of all plots as .png and .pdf. Also sets cache directory.
knitr::opts_chunk$set(
  dev = c("png", "pdf"), fig.keep = "all",
  dev.args = list(pdf = list(encoding = "WinAnsi", useDingbats = FALSE)),
  fig.path = file.path("fig_output/", paste0(gsub("\\.[Rr]md", "/", knitr::current_input()))),
  cache.path = file.path("cache/", paste0(gsub("\\.[Rr]md", "/", knitr::current_input())))
)
```

Load packages
```{r load-packages, message=FALSE, warning=FALSE}
library(dada2); packageVersion("dada2") # sequence filtering and sample inference with dada2
library(ShortRead); packageVersion("ShortRead") # package for loading, displaying, and handling sequences
library(tidyverse); packageVersion("tidyverse") # handy piping and data operations
```

```{r source}
# source all relevant scripting files
# paths.R contains paths to local data and programs
source(file.path("scripts", "paths.R"))
```

Set paths for idemp, cutadapt, and data
````{r test-programs}
# Check by cutadapt version
(cutadapt_messages_1 <- system2(cutadapt_path, args = "--version", stdout = TRUE, stderr = TRUE))
````

| <span> |
| :--- | 
| **NOTE:** idemp relies on having a match in length between the index file and and the barcode sequences. Since the index file usually includes a extra linker basepair (making it 13bp long), you should append the barcode sequences with "N" to make sure each is 13bp long. If you are not sure of the length of index reads, check with the sequencing center. If your index reads are 12bp long, you do NOT need to add an "N". |
| <span> |

Set up file paths in YOUR directory where you want data; 
you do not need to create the subdirectories but they are nice to have
for organizational purposes. 

````{r set-project-dir}
# Set up names of sub directories to stay organized
preprocess_path <- file.path(project_path_OM14, "01_preprocess")
    demultiplex_path <- file.path(preprocess_path, "demultiplexed")
      demultiplex_assigned_path <- file.path(demultiplex_path, "assigned")
      demultiplex_unassigned_path <- file.path(demultiplex_path, "unassigned")
    # filtN_path <- file.path(preprocess_path, "filtN")
    # trimmed_path <- file.path(preprocess_path, "trimmed")
````

# Demultiplex
## Call the demultiplexing script
Demultiplexing splits your reads out into separate files based on the barcodes associated with each sample. 

```{r create-dir-demux}
# Create directory to hold the output from cutadapt
if (!dir.exists(preprocess_path)) dir.create(preprocess_path)
if (!dir.exists(demultiplex_path)) dir.create(demultiplex_path)
if (!dir.exists(demultiplex_assigned_path)) dir.create(demultiplex_assigned_path)
if (!dir.exists(demultiplex_unassigned_path)) dir.create(demultiplex_unassigned_path)
```

```{r demux}
# 2014 barcode length is 12 bases. At error rate of 10% (cutadapt default), there is 1 allowed error.
error_rate_cutadapt <- 0.1

# minimum length to output reads for cutadapt
minimum_length_cutadapt_1 <- 50

flags_cutadapt_1 <- paste("-e", error_rate_cutadapt, "--no-indels", "-m", minimum_length_cutadapt_1, "-g", paste0("file:", barcode_path_OM14), "-o", file.path(demultiplex_assigned_path, "{name}_R1.fastq.gz"), "-p", file.path(demultiplex_assigned_path, "{name}_R2.fastq.gz"), "--untrimmed-output", file.path(demultiplex_unassigned_path, "unassigned_R1.fastq.gz"), "--untrimmed-paired-output", file.path(demultiplex_unassigned_path, "unassigned_R2.fastq.gz"), R1_path_OM14, R2_path_OM14)

cutadapt_messages_1 <- system2(cutadapt_path, args = flags_cutadapt_1, stdout = TRUE, stderr = TRUE)

cutadapt_messages_1 %>% paste(collapse = '\n') %>% cat()
```
```{r list-demux-files}
# Look at output of demultiplexing
list.files(demultiplex_path)
```

## Clean up the demultiplexed output

````{r demux-cleanup}
# Change names of unassignable reads so they are not included in downstream processing
unassigned_1 <- paste("mv", file.path(demultiplex_path, "unknown_R1.fastq.gz"), file.path(demultiplex_path, "Unassigned_reads1.fastq.gz"))
                       
unassigned_2 <- paste("mv", file.path(demultiplex_path, "unknown_R2.fastq.gz"), file.path(demultiplex_path, "Unassigned_reads2.fastq.gz"))
                       
system(unassigned_1)
system(unassigned_2)

# Rename files
# First, generate a vector of desired (shortened) file names

# remove text prior to sample name
R1_names <- gsub(demultiplex_path, "", list.files(demultiplex_path, pattern="R1", full.names = TRUE))

# remove text prior to sample name
R1_names <- gsub("_R1.fastq.gz", "", R1_names)

# put names back together, appending Illumina naming convention, which is the format for later processing with Figaro
R1_names <- paste0(R1_names, "_S0_L001_R1_001.fastq.gz")

# Now, rename files
file.rename(from = list.files(demultiplex_path, pattern="R1", full.names = TRUE), to = file.path(demultiplex_path, R1_names))


# Repeat the above with the reverse reads

R2_names <- gsub(demultiplex_path, "", list.files(demultiplex_path, pattern="R2", full.names = TRUE))

R2_names <- gsub("_R2.fastq.gz", "", R2_names)

R2_names <- paste0(R2_names, "_S0_L001_R2_001.fastq.gz")

file.rename(from = list.files(demultiplex_path, pattern="R2", full.names = TRUE), to = file.path(demultiplex_path, R2_names))
```
