---
---
---
title: "DADA2 ASV table generation from 16S rRNA gene sequences of DNA from Oman samples collected in 2017"
subtitle: "Source file: OM17_dada2_ASV_table_generation.Rmd"
author: "Daniel Nothaft"
date: "`r format(Sys.Date(), '%d %b %Y')`"
output:
  html_document: 
    df_print: paged # omit to disable paged data table output
    css: stylesheet.css # omit if no need for custom stylesheet
    number_sections: yes # change to no for unnumbered sections
    toc: yes # change to no to disable table of contents
    toc_float: true # change to false to keep toc at the top
    toc_depth: 3 # change to specify which headings to include in toc
    code_folding: show # change to hide to hide code by default
editor_options:
  chunk_output_type: inline
---

# Setup

Set knitting options
```{r knitting-options}
# global knitting options for automatic saving of all plots as .png and .pdf. Also sets cache directory.
knitr::opts_chunk$set(
  dev = c("png", "pdf"), fig.keep = "all",
  dev.args = list(pdf = list(encoding = "WinAnsi", useDingbats = FALSE)),
  fig.path = file.path("fig_output/", paste0(gsub("\\.[Rr]md", "/", knitr::current_input()))),
  cache.path = file.path("cache/", paste0(gsub("\\.[Rr]md", "/", knitr::current_input())))
)
```
Load packages
```{r load-packages, message=FALSE, warning=FALSE}
library(dada2); packageVersion("dada2") # sequence filtering and sample inference with dada2
library(ShortRead); packageVersion("ShortRead") # package for loading, displaying, and handling sequences
library(tidyverse); packageVersion("tidyverse") # handy piping and data operations
library(rjson); packageVersion("rjson") # read json output from figaro
```

```{r source}
# source all relevant scripting files
# paths.R contains paths to local data and programs
source(file.path("scripts", "paths.R"))
```
Set directories
````{r set-project-dir}
# Set up names of sub directories to stay organized
preprocess_path <- file.path(project_path_OM17, "01_preprocess")
    demultiplex_path <- file.path(preprocess_path, "demultiplexed")
        demultiplex_assigned_path <- file.path(demultiplex_path, "assigned")
    figaro_output_path <- file.path(preprocess_path, "Figaro_output")
filter_path <- file.path(project_path_OM17, "02_filter") 
````

Now we read in the names of the demultiplexed and assigned fastq files.
```{r read-fastq-files}
fnFs <- sort(list.files(demultiplex_assigned_path, pattern="_R1.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(demultiplex_assigned_path, pattern="_R2.fastq.gz", full.names = TRUE))
```

Count and display reads for input
```{r count-input-reads}
# count forward reads
input_read_count_F <- countLines(demultiplex_assigned_path, pattern="_R1.fastq.gz")/4

# make forward read counts into tbl with sample name
input_read_count_F_tbl <- input_read_count_F %>% as_tibble(rownames = "file_name") %>% rename(reads_F = value) %>% mutate(sample_id = sapply(strsplit(
file_name, "_R1"), `[`, 1))

# repeat for reverse
input_read_count_R <- countLines(demultiplex_assigned_path, pattern="_R2.fastq.gz")/4

input_read_count_R_tbl <- input_read_count_R %>% as_tibble(rownames = "file_name") %>% rename(reads_R = value) %>% mutate(sample_id = sapply(strsplit(
file_name, "_R2"), `[`, 1))

# Join forward and reverse together into one tbl
input_read_count_FR <- input_read_count_F_tbl %>% select(-file_name) %>% left_join(input_read_count_R_tbl %>% select(-file_name), by = "sample_id") %>% select(sample_id, everything())

# print. reads counts are same in forward and reverse, which is what we want.
input_read_count_FR
```
Plot read counts
```{r plot-input-reads, fig.width=10, fig.height=4}
plot_read_count_input <- input_read_count_FR %>% ggplot(aes(
  x = fct_reorder(sample_id, desc(reads_F)),
  y = reads_F
)) +
  geom_bar(stat = "identity") +
  scale_y_continuous(name = "Forward reads") +
  scale_x_discrete(name = "Sample ID") +
  theme_classic(base_size = 9)+
  theme(
    axis.text.x = element_text(angle = 90, vjust = .5, hjust = 1),
    legend.position = "bottom"
  )

plot_read_count_input
```
Extract sample names for later use
```{r extract-original-sample-names}
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.namesF.dots <- sapply(strsplit(basename(fnFs), "_R1"), `[`, 1)

sample.namesR.dots <- sapply(strsplit(basename(fnRs), "_R2"), `[`, 1)

# print first few names
sample.namesF.dots %>% head()
sample.namesR.dots %>% head()
```

# Check primer sequences

```{r primers-515-Y-M13-and-926R, warning=FALSE, message=FALSE}
FWD <- "CCGTAAAACGACGGCCAGTCCGTGYCAGCMGCCGCGGTAA" # 515-Y-M13
REV <- "CCGYCAATTYMTTTRAGTTT" # 926R

(FWD_length <- str_length(FWD))
(REV_length <- str_length(REV))
```

To ensure we have the right primers, and the correct orientation of the primers on the reads, we will verify the presence and orientation of these primers in the data.

```{r Primer-function, warning=FALSE, message=FALSE, tidy=TRUE}
allOrients <- function(primer){ # Create all orientations of the input sequence
  require(Biostrings)
  dna <- DNAString(primer) # The Biostrings works w/ DNAString objects rather than character vectors
  orients <- c(Forward=dna, Complement=complement(dna), Reverse=reverse(dna), RevComp=reverseComplement(dna))
  return(sapply(orients, toString)) # Convert back to character vector
}
FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients
```

We are now ready to count the number of times the primers appear in the forward and reverse read, while considering all possible primer orientations. Identifying and counting the primers on one set of paired end FASTQ files is sufficient, assuming all the files were created using the same library preparation, so we'll just process the first sample.

```{r count-primers}
primerHits <- function(primer, fn) { # Counts number of reads in which the primer is found
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0))
}

rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn=fnFs[[1]]),
      FWD.ReverseReads = sapply(FWD.orients, primerHits, fn=fnRs[[1]]),
      REV.ForwardReads = sapply(REV.orients, primerHits, fn=fnFs[[1]]),
      REV.ReverseReads = sapply(REV.orients, primerHits, fn=fnRs[[1]]))
```

As expected, the FWD primer is found in the forward reads in its forward orientation, and in the reverse reads in its reverse-complement orientation. Similarly, the REV primer is found with its expected orientations. We will remove the primers later when we truncate lower quality reads. So now let's inspect read quality profiles.

# Inspect read quality profiles
It's important to get a feel for the quality of the data that we are using. To do this, we will plot the quality of some of the samples.

*From the dada2 tutorial:*
>In gray-scale is a heat map of the frequency of each quality score at each base position. The median quality score at each position is shown by the green line, and the quartiles of the quality score distribution by the orange lines. The red line shows the scaled proportion of reads that extend to at least that position (this is more useful for other sequencing technologies, as Illumina reads are typically all the same lenghth, hence the flat red line).

````{r plot-qual-R1, fig.width = 7, fig.height = 7}
# Forward reads
# If the number of samples is 20 or less, plot them all, otherwise, just plot 20 randomly selected samples
if( length(fnFs) <= 20) {
  plotQualityProfile(fnFs)
} else {
  rand_samples <- sample(size = 20, 1:length(fnFs)) # grab 20 random samples to plot
  plotQualityProfile(fnFs[rand_samples])
}
````

````{r plot-qual-R2, fig.width = 7, fig.height = 7}
# Reverse reads
# If the number of samples is 20 or less, plot them all, otherwise, just plot 20 randomly selected samples
if( length(fnRs) <= 20) {
  plotQualityProfile(fnRs)
} else {
  rand_samples <- sample(size = 20, 1:length(fnRs)) # grab 20 random samples to plot
  plotQualityProfile(fnRs[rand_samples])
}
````

# Filter parameter optimization with Figaro

Currently, Figaro (v1.1.1) only accepts files with Illumina naming convention, so we need to rename the files.
```{r rename-to-illumina}
# get original file names
R1_names <- gsub(demultiplex_assigned_path , "", list.files(demultiplex_assigned_path , pattern="R1", full.names = TRUE))

# remove suffix
R1_names <- gsub("_R1.fastq.gz", "", R1_names)

# remove dots (Figaro v1.1.1 will throw an error if there dots in the sample name)
R1_names <- R1_names %>% str_replace_all("\\.", "")

# I'm not sure this is necessary, but anyway, I'm adding incrementing Sample digits after S in Illumina style
R1_names_incremented <- vector("character", length(R1_names))
for (i in seq_along(R1_names_incremented )) {
  R1_names_incremented[[i]] <- paste0(R1_names[[i]], "_S", i)
}

# put names back together, appending Illumina naming convention, which is the format for later processing with Figaro
R1_names_incremented <- paste0(R1_names_incremented, "_L001_R1_001.fastq.gz")

# Now, rename files
file.rename(from = list.files(demultiplex_assigned_path , pattern="R1", full.names = TRUE), to = file.path(demultiplex_assigned_path, R1_names_incremented))


# Repeat the above with the reverse reads

R2_names <- gsub(demultiplex_assigned_path , "", list.files(demultiplex_assigned_path , pattern="R2", full.names = TRUE))

R2_names <- gsub("_R2.fastq.gz", "", R2_names)

R2_names <- R2_names %>% str_replace_all("\\.", "")

R2_names_incremented <- vector("character", length(R2_names))
for (i in seq_along(R2_names_incremented )) {
  R2_names_incremented[[i]] <- paste0(R2_names[[i]], "_S", i)
}

R2_names_incremented <- paste0(R2_names_incremented, "_L001_R2_001.fastq.gz")

file.rename(from = list.files(demultiplex_assigned_path , pattern="R2", full.names = TRUE), to = file.path(demultiplex_assigned_path, R2_names_incremented))
```

Check that renaming worked
```{r print-illumina-format-names}
R1_names_incremented %>% head()
```

Create a directory for Figaro output
```{r create-dir-figaro}
# Create directory to hold the output from figaro
if (!dir.exists(figaro_output_path)) dir.create(figaro_output_path)
```
I couldn't get figaro (v1.1.1) to run by system2() in R due to a shell permission denied error, but it worked on the command line.
This chunk generates the exact text I entered in the command line after changing to the figaro directory. I then just read in the figaro output JSON and continue from there.
```{r cd-figaro}
# text to change directory in terminal to figaro 
paste("cd", figaro_path) %>% str_remove("/figaro.py") %>% cat()
```

```{r figaro}
# The length of the amplified sequence target not including primers.
# Kraus et al. (in review) says the 515-Y M13 and 926R primer primer set produce gene fragments of ~400 bp 
# primers total 60 bp
# sequences in 2X250 bp
# I originally tried 340 bp for figaro, but that resulted in low merged read percentage
# EK did not truncate left at all in her original processing
# Thus, I increased to 400
figaro_amplicon_length <- 400

figaro_min_overlap <- 20

figaro_output_file_name <- "Figaro_params_OM17.json"

flags_figaro <- paste("--ampliconLength", figaro_amplicon_length, "--forwardPrimerLength", FWD_length, "--reversePrimerLength", REV_length, "--outputFileName", figaro_output_file_name, "--inputDirectory", demultiplex_assigned_path, "--outputDirectory", figaro_output_path, "--minimumOverlap", figaro_min_overlap)

paste("python3", "figaro.py", flags_figaro) %>% cat()
```

Read in Figaro JSON output, top 5 suggested trim parameters.
```{r read-figaro-json}
# Give the input file name to the function.
figaro_result <- fromJSON(file = file.path(figaro_output_path, figaro_output_file_name))

figaro_result[1:5] %>% paste(collapse = '\n') %>% cat()
```

```{r save-figaro-best-params}
# Save and print Figaro-suggested parameters for dada2::filterAndTrim 

(truncLen_best_figaro <- figaro_result[[1]]$trimPosition)

(maxEE_best_figaro <- figaro_result[[1]]$maxExpectedError)
```

# Filter and trim for quality

## Organize file paths and directories
````{r set-filtering-file-paths}
# set file path for quality-filtered read
filter_path <- file.path(project_path_OM17, "02_filter") 

# Put filtered reads into separate sub-directories for big data workflow

if (!dir.exists(filter_path)) dir.create(filter_path)
    subF_path <- file.path(filter_path, "preprocessed_F") 
    subR_path <- file.path(filter_path, "preprocessed_R") 
if (!dir.exists(subF_path)) dir.create(subF_path)
if (!dir.exists(subR_path)) dir.create(subR_path)
    
# Move R1 and R2 from trimmed to separate forward/reverse sub-directories
# fnFs.Q <- file.path(subF_path,  basename(fnFs)) 
# fnRs.Q <- file.path(subR_path,  basename(fnRs))

fnFs.Q <- file.path(subF_path,  basename(list.files(demultiplex_assigned_path , pattern="R1", full.names = TRUE))) 
fnRs.Q <- file.path(subR_path,  basename(list.files(demultiplex_assigned_path , pattern="R2", full.names = TRUE)))

file.rename(from = list.files(demultiplex_assigned_path , pattern="R1", full.names = TRUE), to = fnFs.Q)
file.rename(from = list.files(demultiplex_assigned_path , pattern="R2", full.names = TRUE), to = fnRs.Q)

# File parsing; create file names and make sure that forward and reverse files match
filtpathF <- file.path(subF_path, "filtered") # files go into preprocessed_F/filtered/
filtpathR <- file.path(subR_path, "filtered") # ...
fastqFs <- sort(list.files(subF_path, pattern="fastq.gz"))
fastqRs <- sort(list.files(subR_path, pattern="fastq.gz"))
if(length(fastqFs) != length(fastqRs)) stop("Forward and reverse files do not match.")
````


Before chosing sequence variants, we want to trim reads where their quality scores begin to drop (the `truncLen` and `truncQ` values) and remove any low-quality reads that are left over after we have finished trimming (the `maxEE` value).

## Filter and trim

````{r filterAndTrim}
# also trimming lengths of primers in this step with trimLeft
filt_out <- filterAndTrim(fwd = file.path(subF_path, fastqFs), filt = file.path(filtpathF, fastqFs), rev = file.path(subR_path, fastqRs), filt.rev = file.path(filtpathR, fastqRs), trimLeft = c(FWD_length, REV_length), truncLen = truncLen_best_figaro, maxEE = maxEE_best_figaro, truncQ = 2, maxN = 0, rm.phix = TRUE, compress = TRUE, verbose = TRUE, multithread = TRUE)

# look at how many reads were kept
head(filt_out)

# summary of samples in filt_out by percentage
filt_out %>% 
  data.frame() %>% 
  mutate(Samples = rownames(.),
         percent_kept = 100*(reads.out/reads.in)) %>%
  select(Samples, everything()) %>%
  summarise(min_remaining = paste0(round(min(percent_kept), 2), "%"), 
            median_remaining = paste0(round(median(percent_kept), 2), "%"),
            mean_remaining = paste0(round(mean(percent_kept), 2), "%"), 
            max_remaining = paste0(round(max(percent_kept), 2), "%"))
````

## Check if primers were removed by filterAndTrim

```{r count-primers-filtered}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn=filtpathF[[1]]),
      FWD.ReverseReads = sapply(FWD.orients, primerHits, fn=filtpathR[[1]]),
      REV.ForwardReads = sapply(REV.orients, primerHits, fn=filtpathF[[1]]),
      REV.ReverseReads = sapply(REV.orients, primerHits, fn=filtpathR[[1]]))
```

The vast majority of primer hits were removed during filtering, but a considerable amount remain. These need to be removed. We will do so with cutadapt.

The primerHits function only scores a hit if the full primer sequence is found in the read. That would be 40 bases for the FWD primer or 20 for the REV primer in this case. The fact that so many full primer sequences remain in the reads after filtering suggest that they are internal (i.e. regular adapters as opposed to anchored adapters in the cutadapt lingo). So, let's use regular adapters. However, to preserve the possibility that some biological reads happen by chance to be the same as the primer sequences, let's increase the minimum overlap in cutadapt from the default of 3 bases to a somewhat more restrictive 5 bases for the REV primers and 7 bases for the FWD primer (because it is longer and thus more likely to have). Since reverse complements are found, let's also enable reverse complements

# Remove primers with cutadapt
## Set up directories for cutadapt
```{r prepare-cutadapt-dirs}
# create directories to put the files trimmed by cutadapt
filt_and_cut_path_F <- file.path(subF_path, "filtered_and_cut")
filt_and_cut_path_R <- file.path(subR_path, "filtered_and_cut") 

# Create directory to hold the output from cutadapt
if (!dir.exists(filt_and_cut_path_F)) dir.create(filt_and_cut_path_F)
if (!dir.exists(filt_and_cut_path_R)) dir.create(filt_and_cut_path_R)

filtfastqFs <- sort(file.path(filtpathF, list.files(filtpathF, pattern="fastq.gz")))

fnFs.cut <- file.path(filt_and_cut_path_F, basename(filtfastqFs))

filtfastqRs <- sort(file.path(filtpathR, list.files(filtpathR, pattern="fastq.gz")))

fnRs.cut <- file.path(filt_and_cut_path_R, basename(filtfastqRs))
```

## Run cutadapt
````{r remove-primers-cutadapt}
# Save the reverse complements of the primers to variables
FWD.RC <- dada2:::rc(FWD)
REV.RC <- dada2:::rc(REV)


# minimum length to output reads for cutadapt
minimum_length_cutadapt <- 50

# minimum overlap for "adapter" (barcode) to be detected by cutadapt
# minimum_overlap_cutadapt_FWD <- 5
# minimum_overlap_cutadapt_REV <- 7
minimum_overlap_cutadapt <- 5

##  Create the cutadapt flags ##
# Trim FWD and the reverse-complement of REV off of R1 (forward reads)
R1.flags <- paste("-g", FWD, "-a", REV.RC)

# Trim REV and the reverse-complement of FWD off of R2 (reverse reads)
R2.flags <- paste("-G", REV, "-A", FWD.RC)


## Run Cutadapt
# default error rate of 10% applied
# https://cutadapt.readthedocs.io/en/stable/guide.html#error-tolerance
# I increased the minimum overlap from the default of 3 to 5 to reduce random matches
# https://cutadapt.readthedocs.io/en/stable/guide.html#minimum-overlap-reducing-random-matches

 # -n 2 required to remove FWD and REV from reads
cutadapt_messages_2_list <- vector("list", length(filtfastqFs)) # set cutadapt message output list
for (i in seq_along(filtfastqFs)) {
     cutadapt_messages_2_list[[i]] <- system2(cutadapt_path, args = c(R1.flags, R2.flags, "--overlap", minimum_overlap_cutadapt, "--minimum-length", minimum_length_cutadapt, "-n", 2, "-o", fnFs.cut[i], "-p", fnRs.cut[i], filtfastqFs[i], filtfastqRs[i]), stdout = TRUE, stderr = TRUE) # input files
}

cutadapt_messages_2_list # print output text
```


Count and display reads for cut
```{r}
# count forward reads
cut_read_count_F <- countLines(filt_and_cut_path_F)/4

# make forward read counts into tbl with sample name
cut_read_count_F_tbl <- cut_read_count_F %>% as_tibble(rownames = "file_name") %>% rename(reads_F = value) %>% mutate(sample_id = sapply(strsplit(file_name, "_S[[:digit:]]"), `[`, 1))

# repeat for reverse
cut_read_count_R <- countLines(filt_and_cut_path_R)/4

cut_read_count_R_tbl <- cut_read_count_R %>% as_tibble(rownames = "file_name") %>% rename(reads_R = value) %>% mutate(sample_id = sapply(strsplit(file_name, "_S[[:digit:]]"), `[`, 1))

# Join forward and reverse together into one tbl
cut_read_count_FR <- cut_read_count_F_tbl %>% select(-file_name) %>% left_join(cut_read_count_R_tbl %>% select(-file_name), by = "sample_id") %>% select(sample_id, everything()) %>% mutate(FR_identical = identical(reads_F, reads_R)) %>% arrange(FR_identical)

# print. reads counts are same in forward and reverse, which is what we want.
cut_read_count_FR
```

Plot read counts
```{r fig.width=10, fig.height=4}
plot_read_count_cut <- cut_read_count_FR %>% ggplot(aes(
  x = fct_reorder(sample_id, desc(reads_F)),
  y = reads_F
)) +
  geom_bar(stat = "identity") +
  scale_y_continuous(name = "Forward reads") +
  scale_x_discrete(name = "Sample ID") +
  theme_classic(base_size = 9)+
  theme(
    axis.text.x = element_text(angle = 90, vjust = .5, hjust = 1),
    legend.position = "bottom"
  )

plot_read_count_cut
```
It looks like, so far, we've kept most of the reads we started with.

## Check primer removal after cutadapt
```{r check-primer-removal-cut-1}
# As a sanity check, we will check for primers in the first cutadapt-ed sample:
    ## should all be zero!
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.cut[[1]]), 
      FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.cut[[1]]), 
      REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.cut[[1]]), 
      REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.cut[[1]]))
```

```{r check-primer-removal-cut-2}
# same for the second sample
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.cut[[2]]), 
      FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.cut[[2]]), 
      REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.cut[[2]]), 
      REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.cut[[2]]))
```

Primers look to be fully removed.

## Plot the quality of the filtered and cut fastq files.

````{r plot-qual-R1-filt-cut, fig.width = 10, fig.height = 7}
# Forward reads
# If the number of samples is 20 or less, plot them all, otherwise, just plot 20 randomly selected samples
if( length(fnFs.cut) <= 20) {
  plotQualityProfile(fnFs.cut)
} else {
  rand_samples <- sample(size = 20, 1:length(fnFs.cut)) # grab 20 random samples to plot
  plotQualityProfile(fnFs.cut[rand_samples])
}
````

````{r plot-qual-R2-filt-cut, fig.width = 10, fig.height = 7}
# Forward reads
# If the number of samples is 20 or less, plot them all, otherwise, just plot 20 randomly selected samples
if( length(fnRs.cut) <= 20) {
  plotQualityProfile(fnRs.cut)
} else {
  rand_samples <- sample(size = 20, 1:length(fnRs.cut)) # grab 20 random samples to plot
  plotQualityProfile(fnRs.cut[rand_samples])
}
````

# Infer sequence variants
In this part of the pipeline, dada2 will learn to distinguish error from biological 
differences using a subset of our data as a training set. After it understands the 
error rates, we will reduce the size of the dataset by combining all identical 
sequence reads into "unique sequences". Then, using the dereplicated data and 
error rates, dada2 will infer the sequence variants in our data. Finally, 
we will merge the corresponding forward and reverse reads to create a list of the 
fully denoised sequences and create a sequence table from the result.

Extract sample names
```{r extract-sample-names}
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.namesF_no_dots <- sapply(strsplit(basename(fnFs.cut), "_S[[:digit:]]"), `[`, 1)
sample.namesR_no_dots <- sapply(strsplit(basename(fnRs.cut), "_S[[:digit:]]"), `[`, 1)

# Double check
if(!identical(sample.namesF_no_dots, sample.namesR_no_dots)) stop("Forward and reverse files do not match.")

# Now, we don't want to add the names without dots (which were only formatted this way to make the files readable for Figaro). Rather, we want the original sample names extracted at the start of this document. Unfortunately, these are not sorted in the same order due to the punctuation change, so we need to do some data wrangling to get the names with dots in the same order.

sample.namesF.dots_tbl <- as_tibble(sample.namesF.dots) %>% rename(name_w_dots = value) %>% mutate(name_no_dots = name_w_dots %>% str_replace_all("\\.", "")) %>% arrange(name_no_dots)

sample.namesF.dots.arranged <- sample.namesF.dots_tbl$name_w_dots

if(!identical(sample.namesF_no_dots, (sample.namesF.dots.arranged %>% str_replace_all("\\.", "")))) stop("Names without dots and original names with dots extracted are different or in different orders. (forward)")

sample.namesR.dots_tbl <- as_tibble(sample.namesR.dots) %>% rename(name_w_dots = value) %>% mutate(name_no_dots = name_w_dots %>% str_replace_all("\\.", "")) %>% arrange(name_no_dots)

sample.namesR.dots.arranged <- sample.namesR.dots_tbl$name_w_dots

if(!identical(sample.namesR_no_dots, (sample.namesR.dots.arranged %>% str_replace_all("\\.", "")))) stop("Names without dots and original names with dots extracted are different or in different orders. (reverse)")

names(fnFs.cut) <- sample.namesF.dots.arranged
names(fnRs.cut) <- sample.namesR.dots.arranged
```

## Learn the error rates

````{r learn-error-rates}
set.seed(100) # set seed to ensure that randomized steps are reproducible

# Learn forward error rates
errF <- learnErrors(fnFs.cut, nbases=5e8, multithread=TRUE)

# Learn reverse error rates
errR <- learnErrors(fnRs.cut, nbases=5e8, multithread=TRUE)
````

### Plot Error Rates
The error rates for each possible transition (A→C, A→G, …) are shown. Points are the observed error rates for each consensus quality score. The black line shows the estimated error rates after convergence of the machine-learning algorithm. The red line shows the error rates expected under the nominal definition of the Q-score. Here, we want to see estimated error rates (black line) that are a good fit to the observed rates (points), and that the error rates drop with increased quality as expected. If these criteria aren't observed, then it may be a good idea to increase the ```nbases``` parameter. This allows the machine learning algorithm to train on a larger portion of your data and may help improve the fit.

````{r plotErrors-f}
plotErrors(errF, nominalQ=TRUE)
```

````{r plotErrors-R}
plotErrors(errR, nominalQ=TRUE)
````

## Dereplication, sequence inference, and merging of paired-end reads

````{r derep-dada}
# make lists to hold the loop output
mergers <- vector("list", length(sample.namesF.dots.arranged))
names(mergers) <- sample.namesF.dots.arranged
ddF <- vector("list", length(sample.namesF.dots.arranged))
names(ddF) <- sample.namesF.dots.arranged
ddR <- vector("list", length(sample.namesF.dots.arranged))
names(ddR) <- sample.namesF.dots.arranged

# For each sample, get a list of merged and denoised sequences
for(sam in sample.namesF.dots.arranged) {
    cat("Processing:", sam, "\n")
    # Dereplicate forward reads
    derepF <- derepFastq(fnFs.cut[[sam]])
    # Infer sequences for forward reads
    ddF[[sam]] <- dada(derepF, err=errF, multithread=TRUE)
    # Dereplicate reverse reads
    derepR <- derepFastq(fnRs.cut[[sam]])
    # Infer sequences for reverse reads
    ddR[[sam]] <- dada(derepR, err=errR, multithread=TRUE)
    # Merge reads together
    merger <- mergePairs(ddF[[sam]], derepF, ddR[[sam]], derepR)
    mergers[[sam]] <- merger
}

rm(derepF); rm(derepR)
````

## Construct sequence table

````{r construct-seqtab}
seqtab <- makeSequenceTable(mergers)
```

```{r write-seqtab}
write_rds(seqtab, path = format(Sys.Date(), "data_output/seqtab_OM17_processed_%Y%m%d_3.rds"), compress = "gz")
```

# Remove Chimeras
Although dada2 has searched for insertion/deletion errors and substitutions, there may still be chimeric
sequences in our dataset (sequences that are derived from forward and reverse sequences from 
two different organisms becoming fused together during PCR and/or sequencing). To identify 
chimeras, we will search for rare sequence variants that can be reconstructed by combining
left-hand and right-hand segments from two more abundant "parent" sequences.

````{r rm-chimeras}
# Remove chimeras
seqtab_nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE)
```

````{r percent-non-chimeric}
# Print percentage of our sequences that were not chimeric.
100*sum(seqtab_nochim)/sum(seqtab)
```

```{r write-seqtab-nochim}
write_rds(seqtab_nochim, path = format(Sys.Date(), "data_output/seqtab_nochim_OM17_processed_%Y%m%d_3.rds"), compress = "gz")
```

# Track reads through pipeline
```{r track}
getN <- function(x) sum(getUniques(x)) # function to grab sequence counts from output objects
# tracking reads by counts
filt_out_track <- filt_out %>%
  data.frame() %>%
  mutate(Sample = gsub("(R1\\_)(.{1,})(\\.fastq\\.gz)","\\2",rownames(.))) %>%
  rename(input = reads.in, filtered = reads.out)
rownames(filt_out_track) <- filt_out_track$Sample

# have to do a few extra string manipulations in this case because of the various sample name changes for Figaro
filt_out_track <- filt_out_track %>% mutate(name_no_dots = sapply(strsplit(Sample, "_S[[:digit:]]"), `[`, 1))

filt_out_track <- filt_out_track %>% left_join(sample.namesF.dots_tbl, by = "name_no_dots")

filt_out_track <- filt_out_track %>% select(-Sample, -name_no_dots) %>% rename(Sample = name_w_dots)

# cutadapted samples
cut_read_count_FR_track <- cut_read_count_FR %>% rename(name_no_dots = sample_id) %>% left_join(sample.namesF.dots_tbl, by = "name_no_dots") %>%  select(-name_no_dots, -FR_identical, -reads_R) %>% rename(Sample = name_w_dots, cut = reads_F)

ddF_track <- data.frame(denoisedF = sapply(ddF[sample.namesF.dots.arranged], getN)) %>%
  mutate(Sample = row.names(.))
ddR_track <- data.frame(denoisedR = sapply(ddR[sample.namesF.dots.arranged], getN)) %>%
  mutate(Sample = row.names(.))
merge_track <- data.frame(merged = sapply(mergers, getN)) %>%
  mutate(Sample = row.names(.))
chim_track <- data.frame(nonchim = rowSums(seqtab_nochim)) %>%
  mutate(Sample = row.names(.))

track <- left_join(filt_out_track, cut_read_count_FR_track) %>%
  left_join(ddF_track, by = "Sample") %>%
  left_join(ddR_track, by = "Sample") %>%
  left_join(merge_track, by = "Sample") %>%
  left_join(chim_track, by = "Sample") %>%
  replace(., is.na(.), 0) %>%
  select(Sample, everything())
row.names(track) <- track$Sample
head(track)

# tracking reads by percentage
track_pct <- track %>% 
  data.frame() %>%
  mutate(Sample = rownames(.),
         filtered_pct = ifelse(filtered == 0, 0, 100 * (filtered/input)),
          cut_pct = ifelse(cut == 0, 0, 100 * (cut/filtered)),
         denoisedF_pct = ifelse(denoisedF == 0, 0, 100 * (denoisedF/cut)),
         denoisedR_pct = ifelse(denoisedR == 0, 0, 100 * (denoisedR/cut)),
         merged_pct = ifelse(merged == 0, 0, 100 * merged/((denoisedF + denoisedR)/2)),
         nonchim_pct = ifelse(nonchim == 0, 0, 100 * (nonchim/merged)),
         total_pct = ifelse(nonchim == 0, 0, 100 * nonchim/input)) %>%
  select(Sample, ends_with("_pct"))

# summary stats of tracked reads averaged across samples
track_pct_avg <- track_pct %>% summarize_at(vars(ends_with("_pct")), 
                                            list(avg = mean))
head(track_pct_avg)


track_pct_med <- track_pct %>% summarize_at(vars(ends_with("_pct")), 
                                            list(avg = stats::median))
head(track_pct_avg)

head(track_pct_med)

# Plotting each sample's reads through the pipeline
track_plot <- track %>% 
  pivot_longer(-Sample, names_to = "Step", values_to = "Reads") %>% 
  mutate(Step = factor(Step, 
                       levels = c("input", "filtered", "cut", "denoisedF", "denoisedR", "merged", "nonchim"))) %>%
  ggplot(aes(x = Step, y = Reads)) +
  geom_line(aes(group = Sample), alpha = 0.2) +
  geom_point(alpha = 0.5, position = position_jitter(width = 0)) + 
  stat_summary(fun.y = median, geom = "line", group = 1, color = "steelblue", size = 1, alpha = 0.5) +
  stat_summary(fun.y = median, geom = "point", group = 1, color = "steelblue", size = 2, alpha = 0.5) +
  stat_summary(fun.data = median_hilow, fun.args = list(conf.int = 0.5), 
               geom = "ribbon", group = 1, fill = "steelblue", alpha = 0.2) +
  geom_label(data = t(track_pct_avg[1:6]) %>% data.frame() %>% 
               rename(Percent = 1) %>%
               mutate(Step = c("filtered", "cut", "denoisedF", "denoisedR", "merged", "nonchim"),
                      Percent = paste(round(Percent, 2), "%")),
             aes(label = Percent), y = 1.1 * max(track[,2])) +
  geom_label(data = track_pct_avg[7] %>% data.frame() %>%
               rename(total = 1),
             aes(label = paste("Total\nRemaining:\n", round(track_pct_avg[1,7], 2), "%")), 
             y = mean(track[,7]), x = 7.6) +
  expand_limits(y = 1.1 * max(track[,2]), x = 8.2) +
  theme_classic()

track_plot
```

